clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java

```diff
@@ -677,13 +789,34 @@ private void updateMetadata(long milliseconds) {
     }
 
     @Override
-    public void subscribe(Pattern pattern, ConsumerRebalanceListener callback) {
-        throw new KafkaException("method not implemented");
+    public void subscribe(Pattern pattern, ConsumerRebalanceListener listener) {
+        maybeThrowInvalidGroupIdException();
+        if (pattern == null || pattern.toString().equals(""))
+            throw new IllegalArgumentException("Topic pattern to subscribe to cannot be " + (pattern == null ?
+                    "null" : "empty"));
+
+        throwIfNoAssignorsConfigured();
+        log.info("Subscribed to pattern: '{}'", pattern);
+        this.subscriptions.subscribe(pattern, listener);
+
+        // this.coordinator.updatePatternSubscription(metadata.fetch());
+        {
+            Cluster cluster = metadata.fetch();
+            // TODO: The following five lines of code are from ConsumerCoordinator.updatePatternSubscription
```

Comment:
Agree, looks to me that we can have this logic as it is, I would only suggest we keep it in a separate `updatePatternSubscription(Cluster cluster)` function

Reply:
Done. Good call.